server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java

```diff
@@ -266,6 +272,12 @@ protected void sendProxyRequest(
       }
     }
 
+    // Since we can't see the request object on the remote side, we can't check whether the remote side actually
```

Comment:
@jon-wei "remote" in this comment means upstream remote (something, that send a request to router), or downstream remote (broker)? Also, in the sentense
```
If the remote node failed to perform an authorization check, will log that on the remote node.
```
apparently the same "remote node" is referred, but in both past tense: "failed" and future tense: "will log", like this code (`clientRequest.setAttribute(AuthConfig.DRUID_AUTHORIZATION_CHECKED, true);`) happens between something else happening on "remote".

I couldn't really understand what is going on here

Reply:
Also, don't understand what does "we can't see the request object on the remote side" mean.

Reply:
"remote" in that comment refers to the proxy forwarding target (the brokers)

> If the remote node failed to perform an authorization check, will log that on the remote node.

Suppose the router forwards a request to a broker, and due to a bug, the broker does not actually perform any authorization checks for that request.

The request will eventually go through the PreResponseAuthorizationCheckFilter on that broker, which will see that no authorization check was performed, and log an error.

> "we can't see the request object on the remote side" 

This refers to how the router has no visibility into whether the broker has set the DRUID_AUTHORIZATION_CHECKED attribute on the request object that the broker is handling.

>     clientRequest.setAttribute(AuthConfig.DRUID_AUTHORIZATION_CHECKED, true);

This is there so that "clientRequest" does not fail the validation checks in PreResponseAuthorizationCheckFilter on the router. The "real" authorization check occurs on the broker that receives a forwarded request.

Reply:
> This refers to how the router has no visibility into whether the broker has set the DRUID_AUTHORIZATION_CHECKED attribute on the request object that the broker is handling.

I don't understand, how this is relevant? If router would magically be able to know, how it would change anything?

> This is there so that "clientRequest" does not fail the validation checks in PreResponseAuthorizationCheckFilter on the router. The "real" authorization check occurs on the broker that receives a forwarded request.

Does this mean that on the query, that arrives to the broker, `DRUID_AUTHORIZATION_CHECKED` is false again, despite this `clientRequest.setAttribute(AuthConfig.DRUID_AUTHORIZATION_CHECKED, true);` line?

Reply:
Is this right, that this line effective just suppresses auth check on the router, because it's done on the broker? If so, why even adding `PreResponseAuthorizationCheckFilter` to the pipeline on router?

Reply:
> I don't understand, how this is relevant? If router would magically be able to know, how it would change anything?

If the router could magically know, it could conceivably have clientRequest's authorization check status match the authorization check status of the proxy request (not that crucial, it would give you some extra information on the router side that there is a potential authorization bug on the forwarding target)

The main point of the comment is just to indicate that the real authorization check occurs at the forwarding destination.

> Does this mean that on the query, that arrives to the broker, DRUID_AUTHORIZATION_CHECKED is false again, despite this clientRequest.setAttribute(AuthConfig.DRUID_AUTHORIZATION_CHECKED, true); line?

Yes, the broker sees the request represented by "proxyRequest", not "clientRequest". Attributes are not transmitted in HTTP requests on the wire anyway, they're only for keeping internal server-side state related to the request.

> Is this right, that this line effective just suppresses auth check on the router, because it's done on the broker? If so, why even adding PreResponseAuthorizationCheckFilter to the pipeline on router?
 
Yes, it's there to suppress auth checks on the router side for forwarded requests. 

There are other endpoints on the router that do need authorization, like `/status` and `/druid/v1/brokers` (which is actually missing authorization checks as of this comment, and should be fixed)